### PR TITLE
[SPARK-42911][PYTHON] Introduce more basic exceptions

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Introduces more basic exceptions.

- ArithmeticException
- ArrayIndexOutOfBoundsException
- DateTimeException
- NumberFormatException
- SparkRuntimeException

### Why are the changes needed?

There are more exceptions that Spark throws but PySpark doesn't capture.

We should introduce more basic exceptions; otherwise we still see `Py4JJavaError` or `SparkConnectGrpcException`.

```py
>>> spark.conf.set("spark.sql.ansi.enabled", True)
>>> spark.sql("select 1/0")
DataFrame[(1 / 0): double]
>>> spark.sql("select 1/0").show()
Traceback (most recent call last):
...
py4j.protocol.Py4JJavaError: An error occurred while calling o44.showString.
: org.apache.spark.SparkArithmeticException: [DIVIDE_BY_ZERO] Division by zero. Use `try_divide` to tolerate divisor being 0 and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error.
== SQL(line 1, position 8) ==
select 1/0
       ^^^

	at org.apache.spark.sql.errors.QueryExecutionErrors$.divideByZeroError(QueryExecutionErrors.scala:225)
... JVM's stacktrace
```

```py
>>> spark.sql("select 1/0").show()
Traceback (most recent call last):
...
pyspark.errors.exceptions.connect.SparkConnectGrpcException: (org.apache.spark.SparkArithmeticException) [DIVIDE_BY_ZERO] Division by zero. Use `try_divide` to tolerate divisor being 0 and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error.
== SQL(line 1, position 8) ==
select 1/0
       ^^^
```

### Does this PR introduce _any_ user-facing change?

The error message is more readable.

```py
>>> spark.sql("select 1/0").show()
Traceback (most recent call last):
...
pyspark.errors.exceptions.captured.ArithmeticException: [DIVIDE_BY_ZERO] Division by zero. Use `try_divide` to tolerate divisor being 0 and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error.
== SQL(line 1, position 8) ==
select 1/0
       ^^^
```

or

```py
>>> spark.sql("select 1/0").show()
Traceback (most recent call last):
...
pyspark.errors.exceptions.connect.ArithmeticException: [DIVIDE_BY_ZERO] Division by zero. Use `try_divide` to tolerate divisor being 0 and return NULL instead. If necessary set "spark.sql.ansi.enabled" to "false" to bypass this error.
== SQL(line 1, position 8) ==
select 1/0
       ^^^
```

### How was this patch tested?

Added the related tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @itholic

File: python/pyspark/errors/exceptions/connect.py [hunk 1]
```diff
@@ -119,13 +135,43 @@ class QueryExecutionException(SparkConnectGrpcException, BaseQueryExecutionExcep
     """
 
 
-class SparkUpgradeException(SparkConnectGrpcException, BaseSparkUpgradeException):
+class PythonException(SparkConnectGrpcException, BasePythonException):
     """
-    Exception thrown because of Spark upgrade from Spark Connect
+    Exceptions thrown from Spark Connect server.
```
Comment (itholic) [inline python/pyspark/errors/exceptions/connect.py:140]: qq: Is `Spark Connect server` and `Spark Connect` different??

Only `PythonException` says it's thrown from Spark Connect "server".

**Review by itholic — APPROVED** [review]
LGTM, left just nit question

**Comment by itholic** [comment]
btw, the example in "Does this PR introduce any user-facing change?" is the same??

**Review by HyukjinKwon — APPROVED** [review]

**Comment by ueshin** [comment]
> the examples in "Does this PR introduce any user-facing change?" are the same??

No, previously we still see `py4j.protocol.Py4JJavaError` or `SparkConnectGrpcException` and now we only see the actual exception classes.

File: python/pyspark/errors/exceptions/connect.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/errors/exceptions/connect.py:140]: The comment is from the previous. We can change it to `Spark Connect` while we are here.

**Comment by itholic** [comment]
Ah I see. One for regular Spark session and the other for remote Spark session.

**Comment by HyukjinKwon** [comment]
@ueshin it has a conflict w/ branch-3.4. would you mind creating a backport PR?

**Comment by ueshin** [comment]
@HyukjinKwon #40547

**Pull request closed without merging**